### PR TITLE
Add DAN to digits_dann_lightn

### PR DESCRIPTION
**Pull request by XianyuanLiu**
### Description
1. Add DAN option in the `digits_dann_lightn/model.py` referring to [ADA](https://github.com/criteo-research/pytorch-ada/blob/ac924cc0780d0d914982950d05b19b4975752c9d/adalib/ada/utils/experimentation.py#L169).
2. Fix the bug raise by #49 in `kale/loaddata/multi_domain.py`. I reverted the changes and added `# type: dict` referring to [mypy issue](https://github.com/python/mypy/issues/922). Is it a good way?

### Status
**Ready**

### Types of changes
- [x] Non-breaking change (fix or new feature that would not break existing functionality).

### CODE REVIEW

**Comment by XianyuanLiu** [comment]
The reason why I push this is I find something wrong when I run DAN on `digits_dann_lightn` and `action_dann_lightn`. The MMD loss `T_mmd` has some values below 0. It will cause `T_total_loss` below 0 because `T_total_loss = T_task_loss + 1 * T_mmd`. Something is wrong, but I can't figure out. Sorry about that. I will raise an issue for this.

**Comment by haipinglu** [comment]
On mypy: the solution mypy suggested is ` _source_by_split: Dict[<type>, <type>] = ..."` and we discussed before string and optional may be the correct types to fill in. I tried but probably used wrong python types so it was not successful. 

Your message on DAN is a bit confusing. Is it ready to merge or it needs further works to fix the problem.

I feel that branches are more convenient when collaborations are needed, saving lots of unnecessary branches or forks or merges.

We could discuss these after Monday's group meeting or in our next individual meeting. I have to get my teaching ready first.

**Comment by XianyuanLiu** [comment]
1. I followed your suggestion to change the code (also follow [pytorch](https://github.com/pytorch/pytorch/blob/916af892b3e7bfa43cab877ee3f2fd7a50f576a8/torch/quantization/fx/quantize.py#L104)).
2. It is ready to merge. There is no DAN option in the `digits_dann_lightn` before. We can only use CDAN and DANN there. I followed the same way in [ADA](https://github.com/criteo-research/pytorch-ada/blob/ac924cc0780d0d914982950d05b19b4975752c9d/adalib/ada/utils/experimentation.py#L170) to add `create_mmd_based` to `model.py`. Thus, we can train and test DAN and JAN on `digits_dann_lightn` now. However, the loss problem is still existing. I push this because I hope others like Shuo can also run DAN in digits to test this problem.
3. I will create a new branch next time. 
4. Sure, we can discuss next Monday. No urgent and I can use CDAN.

**Review by haipinglu — APPROVED** [review]
Thank you for adding DAN to the digits example. That's correct. I did not implement or test DAN earlier. Glad you did now.
